OSDOCS-4491 / OSDOCS-5258: Adding local zones documentation (#53695)
openshift/openshift-docs: modules/rosa-nodes-machine-pools-local-zones.adoc

// Module included in the following assemblies:
//

> // * assemblies/rosa-nodes-machinepools-configuring.adoc
> :_content-type: PROCEDURE

Reviewer: ```suggestion
// * assemblies/rosa-nodes-machinepools-configuring.adoc

:_content-type: PROCEDURE
```

Just doing what everyone else does here. Not sure if this spacing matters, though.

[id="rosa-nodes-machine-pools-local-zones_{context}"]
= Local Zones

Use the following steps to configure Local Zones for machine pools.

[IMPORTANT]
====
AWS Local Zones are supported on Red Hat OpenShift Service on AWS 4.12, and must be enabled by opening a support case with xref:rosa_architecture/rosa-getting-support.html#rosa-getting-support_rosa-getting-support [Red Hat Support].
====
.Prerequisites

> * ROSA is live/GA in the parent region of choice. See the link:https://aws.amazon.com/about-aws/global-infrastructure/localzones/locations/?pg=ln&cp=bn#GA[AWS generally available locations list] to determine the Local Zone available to specific AWS regions.

Reviewer 2: Suggestion: 
s/ROSA is `live/GA` in.../ROSA is generally available in...

The cluster network MTU must always be less than the EC2 MTU to account for the overhead. The specific overhead is determined by your network plugin, for example:

- OVN-Kubernetes: `100 bytes`
- OpenShift SDN: `50 bytes`

The network plugin could provide additional features that may also decrease the MTU. Check the documentation for additional information.

====
* The AWS account has link:https://docs.aws.amazon.com/local-zones/latest/ug/getting-started.html#getting-started-find-local-zone[Local Zones enabled].
* The AWS account has a link:https://docs.aws.amazon.com/local-zones/latest/ug/getting-started.html#getting-started-create-local-zone-subnet[Local Zone subnet] for the same VPC as the cluster.
* The AWS account has a subnet that is associated with a routing table that has a route to a NAT gateway.
* The AWS account has the tag `kubernetes.io/cluster/<infra_id>: shared' on the associated subnet.

.Procedure

> . Create a machine pool on the cluster by running the following ROSA CLI command.

Reviewer: ```suggestion
. Create a machine pool on the cluster by running the following ROSA CLI command:
```

Steps that lead into CLI commands must end with colons.

+
[source,terminal]
----
$ rosa create machinepool -c <cluster-name> -i 
----
+

> . Add the subnet and instance type for the machine pool in ROSA CLI. After several minutes, the cluster will provision the nodes.

Reviewer: Same comment as above.

Author: I do not know what "as above" is referring to. Can you please clarify?

Reviewer: Hm. This isn't actually entering the terminal on its own step, so I guess the colon doesn't apply. You've entered the interface by way of the previous step. 

I'm more accustomed to seeing this situation formatted like so: https://docs.openshift.com/container-platform/4.12/installing/installing_aws/installing-aws-network-customizations.html#installation-initializing_installing-aws-network-customizations, and that is what my gut says it should be like. You avoid the a11y issues of callouts that way, too.

Reviewer: I'd also note that the example mixes user-replaceable values with concrete values that seem to really be user-replaceable, e.g. callout 5 vs. callout 7.

+
[source, terminal]
----
I: Enabling interactive mode <1>
? Machine pool name: xx-lz-xx <2>
? Create multi-AZ machine pool: No <3>
? Select subnet for a single AZ machine pool (optional): Yes <4>
? Subnet ID: subnet-<a> (region-info) <5>
? Enable autoscaling (optional): No <6>
? Replicas: 2 <7>
I: Fetching instance types <8>
----

> +
> 

Reviewer: ```suggestion
```
Verify that this doesn't work better for you. Callouts typically follow code blocks immediately.

Author: Removing this changes the numbering structure, so it needs to remain.

<1> Enables interactive mode.
<2> Names the machine pool. This is limited to alphanumeric and a maximum length of 30 characters.
<3> Set this option to no.
<4> Set this option to yes.
<5> Selects a subnet ID from the list.
<6> Select yes to enable autoscaling or no to disable autoscaling.
<7> Selects the number of machines for the machine pool. This number can be anywhere from 1 - 180.
<8> Selects an instance type from the list. Only instance types that are supported in the selected Local Zone will appear.

> +

Reviewer: ```suggestion
```

> +
> 
> <1> Enables interactive mode.
> <2> Names the machine pool. This is limited to alphanumeric and a maximum length of 30 characters.
> <3> Set this option to no.
> <4> Set this option to yes.
> <5> Selects a subnet ID from the list.
> <6> Select yes to enable autoscaling or no to disable autoscaling.
> <7> Selects the number of machines for the machine pool. This number can be anywhere from 1 - 180.
> <8> Selects an instance type from the list. Only instance types that are supported in the selected Local Zone will appear.
> +

Reviewer: ```suggestion
<1> Enables interactive mode.
<2> Names the machine pool. This is limited to alphanumeric and a maximum length of 30 characters.
<3> Set this option to no.
<4> Set this option to yes.
<5> Selects a subnet ID from the list.
<6> Select yes to enable autoscaling or no to disable autoscaling.
<7> Selects the number of machines for the machine pool. This number can be anywhere from 1 - 180.
<8> Selects an instance type from the list. Only instance types that are supported in the selected Local Zone will appear.
```

Standard adoc callout formatt has callouts immediately following a block.

> . Provide the subnet ID to provision the machine pool in the Local Zone.

Reviewer: Is this part of the previous step, or a command-line program flow? Not sure where to provide it, otherwise.

Author: No, it is not part of the previous step, and no, it is not a command-line program flow. This is exactly how it needs to be written.

Reviewer: 🤷 I think that it's confusing, but I'm just one reviewer. The step doesn't say where I'm providing the ID.

// Module included in the following assemblies:
//
// * assemblies/rosa-nodes-machinepools-configuring.adoc

> :_content-type: CONCEPT

Reviewer: ```suggestion
:_content-type: PROCEDURE
```

[id="rosa-nodes-machine-pools-local-zones_{context}"]

> = Local Zones

Reviewer: This title doesn't seem like a procedure title to me.

```suggestion
= Configuring Local Zones for machine pools
```

Uninspired suggestion. :)

Author: I like your suggestion, but I'm going to keep the title as-is due to feedback from SMEs and PMs.

Author: I have made the suggested changes.

> Use the following steps to configure Local Zones for machine pools.

Reviewer: Self-referential.

[IMPORTANT]
====
AWS Local Zones are supported on Red Hat OpenShift Service on AWS 4.12. See the link:https://access.redhat.com/articles/6989889[Red Hat Knowledgebase article] for information on how to enable Local Zones. 
====
.Prerequisites

* ROSA is generally available in the parent region of choice. See the link:https://aws.amazon.com/about-aws/global-infrastructure/localzones/locations/?pg=ln&cp=bn#GA[AWS generally available locations list] to determine the Local Zone available to specific AWS regions.
* The ROSA cluster was initially built in an existing Amazon VPC (BYO-VPC).
* The maximum transmission unit (MTU) for the ROSA cluster is set at 1200.
+
[IMPORTANT]
====
Generally, the Maximum Transmission Unit (MTU) between an Amazon EC2 instance in a Local Zone and an Amazon EC2 instance in the Region is 1300. See link:https://docs.aws.amazon.com/local-zones/latest/ug/how-local-zones-work.html[How Local Zones work] in the AWS documentation. 
The cluster network MTU must be always less than the EC2 MTU to account for the overhead. The specific overhead is determined by your network plugin, for example:

- OVN-Kubernetes: `100 bytes`
- OpenShift SDN: `50 bytes`

> The network plugin could provide additional features, like IPsec, that also must be decreased the MTU. Check the documentation for additional information.

Author: Done, and fixed "must be decreased" too.

====
* The AWS account has link:https://docs.aws.amazon.com/local-zones/latest/ug/getting-started.html#getting-started-find-local-zone[Local Zones enabled].
* The AWS account has a link:https://docs.aws.amazon.com/local-zones/latest/ug/getting-started.html#getting-started-create-local-zone-subnet[Local Zone subnet] for the same VPC as the cluster.
* The AWS account has a subnet that is associated with a routing table that has a route to a NAT gateway.
* The AWS account has the tag `kubernetes.io/cluster/<infra_id>: shared' on the associated subnet.

> 

Reviewer: Consider deleting repeated line breaks throughout, but that's not a necessity.

.Procedure

> . Create a machine pool on the cluster in ROSA CLI.
> +
> [source,terminal]
> ----
> $ rosa create machinepool -c <cluster-name> -i 
> ----

Reviewer: ```suggestion
. Create a machine pool on the cluster by running the following ROSA CLI command:
+
[source,terminal]
----
$ rosa create machinepool -c <cluster-name> -i 
----
```

https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output

Same elsewhere.

// Module included in the following assemblies:
//
// * assemblies/rosa-nodes-machinepools-configuring.adoc
:_content-type: CONCEPT
[id="rosa-nodes-machine-pools-local-zones_{context}"]
= Local Zones

Use the following steps to configure Local Zones for machine pools.

[IMPORTANT]
====

> AWS Local Zones are supported on Red Hat OpenShift Service on AWS 4.12, and must be enabled by opening a support case with xref:rosa_architecture/rosa-getting-support.html#rosa-getting-support_rosa-getting-support [Red Hat Support].

Reviewer 3: The link for 'Red Hat Support' can't reach

Author: Because this is a preview, the links won't work as intended. The link will work once it's live.

Reviewer 2: I think it's the the missing ../ and space between the xref and the title that's causing the link to fail. Give this a try:

xref:../rosa_architecture/rosa-getting-support.html#rosa-getting-support_rosa-getting-support[Red Hat Support].

- OVN-Kubernetes: `100 bytes`
- OpenShift SDN: `50 bytes`

The network plugin could provide additional features, like IPsec, that also must be decreased the MTU. Check the documentation for additional information.

====
* The AWS account has link:https://docs.aws.amazon.com/local-zones/latest/ug/getting-started.html#getting-started-find-local-zone[Local Zones enabled].
* The AWS account has a link:https://docs.aws.amazon.com/local-zones/latest/ug/getting-started.html#getting-started-create-local-zone-subnet[Local Zone subnet] for the same VPC as the cluster.

.Procedure

. Create a machine pool on the cluster in ROSA CLI.
+
[source,terminal]
----
$ rosa create machinepool -c <cluster-name> -i 
----
+
. Add the subnet and instance type for the machine pool in ROSA CLI. After several minutes, the cluster will provision the nodes.
+
[source, terminal]
----
I: Enabling interactive mode <1>
? Machine pool name: xx-lz-xx <2>
? Create multi-AZ machine pool: No <3>
? Select subnet for a single AZ machine pool (optional): Yes <4>
? Subnet ID: subnet-<a> (region-info) <5>

> ? Enable autoscaling (optional): No <6>

Reviewer 3: There is no requirement to set it to ‘No’. 'Yes' or 'No' are ok for localzone

? Replicas: 2 <7>
I: Fetching instance types <8>
----
+

<1> Enables interactive mode.
<2> Names the machine pool. This is limited to alphanumeric and a maximum length of 30 characters.
<3> Set this option to no.
<4> Set this option to yes.

> <5> Selects a subnet ID from the list. Associate the subnet with a routing table that has a route to a NAT gateway.

Reviewer 3: This tip should show before "create machinepool".

> Associate the subnet with a routing table that has a route to a NAT gateway

<6> Set this option to no.
<7> Selects the number of machines for the machine pool. This number can be anywhere from 1 - 180.
<8> Selects an instance type from the list. Only instance types that are supported in the selected Local Zone will appear.
+

> . Add the `kubernetes.io/cluster/<infra_id>: shared` tag to the subnet.

Reviewer 3: The step should show before "create machinepool".

. Provide the subnet ID to provision the machine pool in the Local Zone.

See the link:https://aws.amazon.com/about-aws/global-infrastructure/localzones/locations/[AWS Local Zones locations] list on AWS for generally available and announced AWS Local Zone locations.

[NOTE]
====

> Local machine pools can be added to single Availability Zone clusters.

Reviewer 3: It allows creating a single zone machine pool with subnet(created on local zone) to a multi-az cluster
So the note is not right

Author: I've gotten conflicting information on the single zone / multi-az issue. @oriAdler @arendej , can y'all please weigh in?
